💬 guidelines on feedback and running (#39)
exercism/automated-analysis: docs/guidelines.md

# Feedback guidelines

The goal of a language track on Exercism is to give people a way to achieve a
high level of fluency at a low level of proficiency. We're aiming for fluency
in the syntax, idioms, and the standard library of the language. You can read
more about the [goal of exercism here](https://github.com/exercism/docs/blob/master/about/goal-of-exercism.md).

## Definitions

In the following paragraphs, keywords such as **MUST**, **SHOULD**, **COULD**
are to be interpreted as in [RFC2119](https://www.ietf.org/rfc/rfc2119.txt);
given that we recognise the following four output states and their restrictions:

> - `approve_as_optimal`: **SHOULD** be a best solution, **SHOULD** be without

Reviewer: The website will drop any comments in this scenario. Does that mean it should be `**MUST** be without comment` or because the website handles, SHOULD is ok?

Reviewer: In what scenarios would this not be a best solution? The website can't guard for this, so it should maybe be `**MUST**`?

Author: Unfortunately no. It **SHOULD** be, because in fact you and I for example do structural testing. We can not _guarantee_ that all `approve_as_optimal` are _thé_ best solutions there are. We can only add as many assertions as possible that it's likely and probably.

**MUST** here would mean that someone would be able to blindly copy anything that's `approve_as_optimal` and it would _never_ include anything that's less than optimal. The fact that the website can't guard is exactly why it shouldn't be MUST.

> The website will drop any comments in this scenario

You're correct: because the website does not _break_ if you do send it, it's **SHOULD**.

Reviewer 2: > We can not guarantee that all approve_as_optimal are thé best solutions there are

Which side is that? The analyzer must guarantee that (I think). Or is that the website which is not able to guarantee it?

Reviewer 2: The more I think about it the more I am sure this should be: `**MUST** be a best solution`.

If we can guarantee it or not doesn't matter. We would consider it a bug if it is not a "best solution" and fix it.

**SHOULD** implies that there might be valid reasons to use `approve_as_optimal` although the solution is not optimal.

Definition:
```
3. SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.
```

Author: Yes. I've changed my mind now. I agree with both @tehsphinx @iHiD.

Let's make this **MUST** be a best solution. The other terms seem fine.

standard** and become idiomatic use. Example: **Ruby** uses 2 space identation.

Some rules are language features, even if they are not documented well. These
language features are part of "idiomatic rules" and not stylistic choices.
Example: **Ruby**'s MRI treats variables named `_` differently.

Finally there are rules that are pure preferences, even though they might be
adopted by large bodies such as organisations and corporations. These rules
are usually part of _competing_ standards. Exercism does not favour one over
another. Example: **TypeScript** has a linter `tslint` (or `eslint` + plugin)
which is maintained by a company that is not Microsoft. It competes with other
linters such as `xo`. Most of the rules are not language features or idiomatic
rules, and therefore stylistic choices.

## Conditions and outcome

For feedback generated by automated mentoring:

- it **SHOULD** `disapprove_with_comment` when someone isn't using one of the
  subjects the exercise is supposed to teach,
- each `disapprove_with_comment` **MUST** have at least one comment,
- each `disapprove_with_comment` **SHOULD** steer a student towards an
  `approve_as_optimal`,

> - a student **SHOULD NOT** get a `refer_to_mentor` in a newer iteration, if

Reviewer: I think this should be worded more along the lines of "The analyzer should aim to teach a pathway that means a student never gets a refer_to_mentor after a `disapprove_with_comment`. The analyzers are currently stateless though, so they cannot guard against whether there has previously been a `disapprove_with_comment` or not

  they've had a `disapprove_with_comment` in a previous iteration,
- a `refer_to_mentor` **COULD** have one or more comments which will be given to
  the mentor.

> There is no minimum or maximum amount of :speech_balloon: comments. However,

Reviewer: Maybe :)

I'd maybe say "It is currently undecided as to whether there is a maximum number of comments. Our recommendation is currently to aim for one comment per iteration" (because that's my gut instinct, although far from certain).

Author: Agreed!

each comment **SHOULD** be aiding the student towards an approvable solution.

## Approvability

Given the above, and to re-iterate that we focus on language fluency:

### Incorrect Naming or Casing

In general, if it's a _language feature_ that will be caught by a compiler or
parser or based on official rules (which means there is a dependency on
correctness in tools), you should `disapprove_with_comment`, preferably linking
to the official rules. 

> > - :-1: dissapprove if there are official guidelines on naming conventions

Reviewer 3: ```suggestion
> - :-1: disapprove if there are official guidelines on naming conventions
```

> - :speech_balloon: leave a comment if there is something noteworthy
> - :no_bell: if its a _stylistic_ preference, **and** there are _competing
>   standards_, do not remark at all. Since there are competing standards,
>   they're all preferences.
> - :speech_balloon: if it's a _stylistic_ preference, **and** there is _one
>   clear standard_, comment on it. These rules enforce idiomatic code.
> - :question: If it's a _stylistic_ preference, **and** there is no clear
>   standard, but most to all non-hobyist have adopted the same style, this
>   might be idiomatic. Comment at your discretion.

#### Examples

- **Ruby** has a language feature where `_` is treated differently,
  - :-1: if a student uses `_` for a variable name, but then uses it.
- **Ruby** recognises `constants` only if they start with a **C**apital Letter,
  - :-1: if a student uses `snake_case` for a `class` name

> - **Ruby** has _defacto_ standards on `cAsInG` and `name-ing`,

Reviewer 3: ```suggestion
- **Ruby** has _de facto_ standards on `cAsInG` and `name-ing`,
```

>   - :no_bell: if a student uses `camelCase` for a variable, leave it be.

Reviewer: Why is this 🔕? That seems to contradict the next line.

Author: You're right. That is not the case for `Ruby`.

  - :speech_balloon: you **SHOULD** guide students that `snake_case` is to be
    expected by most IDEs and highlighting on exercism in code blocks.
- **JavaScript** IDEs highlight variables which are not used, except for those
  prefixed with an underscore (`_`).
  - :speech_balloon: note that this behaviour exist so it might help them to use
    a different naming strategy. They might think that prefixing with `_` means 
    `private`, which is not the case in JavaScript.

> - **TypeScript** has a _defacto_ standard lint tool provided by Palantir,

Reviewer 3: ```suggestion
- **TypeScript** has a _de facto_ standard lint tool provided by Palantir,
```

# Feedback guidelines

The goal of a language track on Exercism is to give people a way to achieve a
high level of fluency at a low level of proficiency. We're aiming for fluency
in the syntax, idioms, and the standard library of the language. You can read
more about the [goal of exercism here](https://github.com/exercism/docs/blob/master/about/goal-of-exercism.md).

## Definitions

In the following paragraphs, keywords such as **MUST**, **SHOULD**, **COULD**
are to be interpreted as in [RFC2119](https://www.ietf.org/rfc/rfc2119.txt);
given that we recognise the following four output states and their restrictions:

> - `approve_as_optimal`: **COULD** be a best solution, **SHOULD** be without comment

Reviewer 2: Isn't `having a best solution` a requirement or at least a recommendation of `approve_as_optimal`?

Reviewer 2: If `approve_as_optimal` can have a comment, what is the difference to `approve_with_comment`?

Reviewer: `approve_as_optimal` is for solutions where no comment about an improved is required.

Author: So @tehsphinx if you take @iHiD comment and the one that was made about _random tips_, we might want to later use `approve_as_optimal` still but also add a comment.

The reason I used `COULD` and `SHOULD` here is that if I used `MUST`, technically a test framework could blow up the analyzer if a tip is added, or if the approved solution is not technically a "best" solution.

Reviewer 4: @SleeplessByte I understand the reasoning, but to me `**COULD** be a best solution` still feels a bit odd. I would at least have expected **SHOULD** here, but maybe that's me?

Author: I'm indifferent. I'll change it so it's clearer!

Author: Done!

Reviewer 4: Nice!

Reviewer 4: @SleeplessByte I'm not seeing the change though :)

- `approve_with_comment`: **COULD** be a best solution, **MUST** be with comment
- `disapprove_with_comment`: **MUST** be with comment
- `refer_to_mentor`: **COULD** be with comment

## Conditions and outcome

For feedback generated by automated mentoring:

> - it **SHOULD** `disapprove_with_comment` when someone isn't using one of the 

Reviewer 5: trailing white space

  subjects the exercise is supposed to teach,

> - each `disapprove_with_comment` **MUST** has at least one comment,

Reviewer 3: ```suggestion
- each `disapprove_with_comment` **MUST** have at least one comment,
```

- each `disapprove_with_comment` **SHOULD** steer a student towards an `approve_`,
- a student **SHOULD NOT** get a `refer_to_mentor` in a newer iteration, if they've
  had a `disapprove_with_comment` in a previous iteration,

> - a `refer_to_mentor` **COULD** have one or more comments which will be given to

Reviewer 4: So these comments will only be passed to the mentor, not the student?

Author: TBD -- currently no comments will be passed to anyone 🚫. This will be up to the product team ;)

I wrote it like this because an earlier comment by @iHiD indicated that `refer_to_mentor` would refer the comments to the mentor! Such as the "more than 3 components interpolation" string.

Personally I think it would be _really_ helpful for mentors to understand _why_ it was referred to them.

Reviewer 4: Agreed, although in many cases it will be: "because I (the analyzer) don't understand the code" :)

Author: Definitely! But the great thing is that we _could_ collect _what_ it doesn't understand and then iterate. Although that might be better in `analysis.out`

Reviewer: I'm happy with this. I have lots of cases where I actively refer to mentor knowing what is weird but not how to fix it. I've acutally decided in lots of places not to tell the student what is wrong even though I could work it out, because I think it'll be better with a human conversation.

  the mentor.

> There is no minumum or maximum amount of :speech_balloon: comments. However, each

Reviewer 5: ~minumum~ minimum

comment **SHOULD** be aiding the student towards an approvable solution.

## Approvability

Given the above, and to re-iterate that we focus on language fluency:

### Incorrect Naming or Casing

In general, if it's a _language feature_ that will be caught by a compiler or parser

> or based on official rules (which means there is a dependency on correctness in 

Reviewer 5: trailing white space

tools), you should `disapprove_with_comment`, preferably linking to the official

> rules. If it's a _stylistic_ choise, or there are _competing standards_, do not

Reviewer 5: ~choise~ choice

> remark at all.

Reviewer 3: Suggestion for the last sentence:

If it's a _stylistic_ choice, **and** there are _competing standards_, do not remark at all. **If there's one clear standard, even if not official, consider commenting on it.**

(Marked changes in bold. And feel free to tweak the language a bit.)

> - :-1: dissapprove if there are official *strict* guidelines on naming conventions
> - :+1: approve if there are only _defacto_ standard(s) and everything else is fine.
> - :speech_balloon: leave a comment if there is something noteworthy
> - :no_bell: ignore stylistic issues if it's not integral to the language

#### Examples

- **Ruby** has a language feature where `_` is treated differently,
  - :-1: If a student uses `_` for a variable name, but then uses it.

> - **Ruby** reconises `constants` only if they start with a **C**apital Letter,

Reviewer 5: ~reconises~ recognizes

> - **Ruby** reconises `constants` only if they start with a **C**apital Letter,

Reviewer 5: ~reconises~ recognises

  - :-1: If a student uses `snake_case` for a `class` name
- **Ruby** has _defacto_ standards on `cAsInG` and `name-ing`,
  - :no_bell: If a student uses `camelCase` for a variable, leave it be.

>   - :speech_balloon: You may guide students that `snake_case` is to be expected by

Reviewer 3: ```suggestion
  - :speech_balloon: You should guide students that `snake_case` is to be expected by
```

Should we **encourage** analyzers to add a comment if the solution violates a widely accepted standard (even if not officially enforced)?

Author: I like that.

>     most IDE's and highlighting on exercism in code blocks for variables.

Reviewer 5: I believe this is meant to be plural and not possessive.  ~IDE's~ IDEs

Author: Yes it is!

- **JavaScript** IDEs highlight variables which are not used, except for those
  prefixed with an underscore (`_`).
  - :speech_balloon: Note that this behaviour exist so it might help them to use a
    different naming strategy. They might think that prefixing with `_` means 
    `private`, which is not the case in JavaScript.
- **TypeScript** has a _defacto_ standard lint tool provided by Palantir,
  - :no_bell: If a student does not follow these rules as the lint tool is not 
    official. In fact, there are multiple linters out there, with mutually exclusive
    rules.

> - **Go** has very strict rules around naming and other linting.

Reviewer 2: ```suggestion
- **Go** has very strict rules around formatting and linting.
```
Or change this, then we can keep `gofmt` below.

Author: Nah, I liked the other one 👍 . We can of course add both 📦

>   - :-1: If the student does not follow these rules (e.g. has not applied `gofmt`)

Reviewer 2: ```suggestion
  - :-1: If the student does not follow these rules (e.g. has not applied `golint`)
```
`gofmt` doesn't do naming or casing
